### PR TITLE
Optimize rendering of text for timers.

### PR DESCRIPTION
**Pull request by danielfenner**
Previously the fitting of the text into the boxes for the timers took too long. Root cause for this is a fairly complicated computation for determining the width of a rendered text. This seems to be used inside the method to elide text as well.
This PR replaces the QFontMetrics::elidedText and
QFontMetrics::horizontalAdvance by a simplified, less accurate method. We compute a lookup table storing the rendered width of all ascii characters. The width is computed as the sum of the width of all characters in a string (plus a heuristic to obtain an upper bound of the actual width). The method to elide text is is implemented on basis of this simplified width calculation.
Compare https://github.com/google/orbit/issues/4626.

### CODE REVIEW

File: src/OrbitGl/QtTextRenderer.cpp [hunk 1]
```diff
@@ -214,38 +237,87 @@ float QtTextRenderer::GetStringWidth(const char* text, uint32_t font_size) {
...
+  const int horizontal_advance = MaximumHeuristic(width, text.length(), font_size);
+  return viewport_->ScreenToWorld(Vec2i(horizontal_advance, 0))[0];
+}
+
+QString QtTextRenderer::ElideText(const QString& text, int max_width,
+                                  const CharacterWidthLookup& lookup, uint32_t font_size) {
+  int width_lookup = 0;
+  int characters = 0;
+  while (characters < text.length()) {
+    const int next_char_width = lookup[text[characters].toLatin1()];
```
Comment (florian-kuebler) [inline src/OrbitGl/QtTextRenderer.cpp:299]: Same here.

**Review by florian-kuebler — APPROVED** [review]
LGTM. Thank you!

**Review by beckerhe — APPROVED** [review]
Awesome. Otherwise LGTM

File: src/OrbitGl/QtTextRenderer.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (danielfenner) [reply src/OrbitGl/QtTextRenderer.cpp:298]: see above

File: src/OrbitGl/QtTextRenderer.cpp [hunk 2]
```diff
@@ -29,6 +31,18 @@ namespace orbit_gl {
...
 
+// Qt offers a QFontMetrics::horizontalAdvance to determine the width of a rendered string. This
+// method is fairly slow. For rendering the text in the timers we therefore use a different method:
+// We compute a lookup table storing the rendered width of all the characters and sum over all the
+// characters in a string (compare GetCharacterWidthLookup, GetStringWidthFast below). The result is
+// consistently a bit shorter than the correct result provided by QFontMetrics::horizontalAdvance.
+// Applying the heuristic below to the result from the lookup reliably yields a fairly tight upper
+// bound for the true width of the rendered string. Don't try to make sense of the formula - it is
+// just a line fitted to example data.
+inline int MaximumHeuristic(int width, int length, int font_size) {
```
[the remark this answers is not in the document]
Comment (danielfenner) [reply src/OrbitGl/QtTextRenderer.cpp]: done

File: src/OrbitGl/QtTextRenderer.cpp [hunk 3]
```diff
@@ -214,38 +237,87 @@ float QtTextRenderer::GetStringWidth(const char* text, uint32_t font_size) {
...
+    return lut;
+  }
+  return it->second;
+}
+
+float QtTextRenderer::GetStringWidthFast(const QString& text, const CharacterWidthLookup& lookup,
+                                         uint32_t font_size) {
+  int width = 0;
+  for (const QChar& c : text) {
+    width += lookup[c.toLatin1()];
```
[the remark this answers is not in the document]
Comment (beckerhe) [reply src/OrbitGl/QtTextRenderer.cpp:287]: Depending on the implementation of char's range is either `-256...255` or `0...511` - but `lookup` has size 256.

[the remark this answers is not in the document]
Comment (danielfenner) [reply src/OrbitGl/QtTextRenderer.cpp:287]: Oh, yeah, not quiet but there is a problem. 
  ```
  static_assert(sizeof(char) == 1);
  static_assert(std::numeric_limits<char>::min() == -128);
  static_assert(std::numeric_limits<char>::max() == 127);
```
So `char` is one byte as expected but it is signed.

[the remark this answers is not in the document]
Comment (beckerhe) [reply src/OrbitGl/QtTextRenderer.cpp:287]: Oh, right. I got the numbers wrong. But you're right. Signedness is a problem.

**Pull request merged**